### PR TITLE
Implement DictRead hint

### LINKED ISSUE #289: DictRead
Reads a value from a dictionary at a specified key.

Reference: https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/dict.cairo#L25

### PR DESCRIPTION
**Pull request by har777**
Fixes #289

### CODE REVIEW

**Comment by har777** [comment]
Moving to draft till integration tests added.

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 1]
```diff
@@ -10,6 +10,122 @@ import (
...
+			if err != nil {
+				return err
+			}
+			dictionaryManager, ok := ctx.ScopeManager.GetZeroDictionaryManager()
+			if !ok {
+				return fmt.Errorf("__dict_manager not in scope")
+			}
+
+			//> dict_tracker.current_ptr += ids.DictAccess.SIZE
+			err = dictionaryManager.IncrementFreeOffset(*dictPtr, 3)
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:90]: are you sure that incrementing the free offset is enough? dont we need to also write the entry in the dict memory segment?

Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:90]: lgtm for the rest

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go [hunk 2]
```diff
@@ -4,11 +4,70 @@ import (
...
+					err := ctx.runnerContext.ScopeManager.AssignVariable("__dict_manager", dictionaryManager)
+					if err != nil {
+						t.Fatal(err)
+					}
+					defaultValueMv := memory.MemoryValueFromInt(12345)
+					dictionaryManager.NewDefaultDictionary(ctx.vm, defaultValueMv)
+					return newDictReadHint(ctx.operanders["dict_ptr"], ctx.operanders["key"], ctx.operanders["value"])
+				},
+				check: varValueEquals("value", feltUint64(12345)),
+			},
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_dictionaries_test.go:69]: lgtm

**Review by TAdev0 — COMMENTED** [review]
just one comment about the incrementation of the free memory pointer, with or without writing the read entry in the dict memory segment?

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 1 above)
Comment (har777) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:90]: Yeah thats all the python code does!

Memory is updated as part of the main `dict_read` cairo function: https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/dict.cairo#L22

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 3]
```diff
@@ -59,6 +59,73 @@ func createDefaultDictNewHinter(resolver hintReferenceResolver) (hinter.Hinter,
...
 
+// DictRead hint reads the result of given key from given dictionary
+// and writes it to value
+//
+// `newDictReadHint` takes 3 operanders as argument
+//   - `dictPtr` variable will be pointer to the dictionary to read from
+//   - `key` variable will be the key read from the dictionary
+//   - `value` variable will be the variable to which the result of
+//     the `key` read from the dictionary pointed to by `dictPtr`
+//     will be written to
```
Comment (rodrigo-pino) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:70]: I think this description might be simpler.

`value` variable will hold the value of the dictionary stored at `key`

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 4]
```diff
@@ -59,6 +59,73 @@ func createDefaultDictNewHinter(resolver hintReferenceResolver) (hinter.Hinter,
 	return newDefaultDictNewHint(defaultValue), nil
 }
 
+// DictRead hint reads the result of given key from given dictionary
+// and writes it to value
```
[the remark this answers is not in the document]
Comment (har777) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:63]: I'll try to format it better. Technically its neither returning a value nor is it writing anything to the segment. All of that is done by the cairo `dict_read` function which calls this hint as part of its code.

**Review by TAdev0 — APPROVED** [review]

**Pull request merged**